keps/sig-node/20190717-seccomp-ga.md

```diff
@@ -0,0 +1,378 @@
+the key, this is entirely useable on any production GA cluster.
+
+There have been multiple attempts to [change the default
+profile](https://github.com/kubernetes/kubernetes/issues/39845) or [formally spec the Kubernetes
+seccomp profile](https://github.com/kubernetes/kubernetes/issues/39128), but both efforts were
+abandoned due to friction and lack of investment.
+
+Despite the `alpha` label, I think this feature needs to be treated as GA, and we're doing our users
+a disservice by leaving it in this weird limbo state. As much as I would like to see seccomp support
+fully fleshed out, if we block GA on those enhancements we will remain stuck in the current state
+indefinitely. Therefore, I'm proposing we do the absolute minimum to clean up the current
+implementation all accurately declare the feature "GA". Future enhancements can follow the standard
+alpha -> beta -> GA feature process.
+
+_NOTE: AppArmor is in a very similar state, but with some subtle differences. Promoting AppArmor to
+GA will be covered by a separate KEP._
+
+### Goals
+
+- Declare seccomp GA
+- Fully document and formally spec the feature support
+- Add equivalent API fields to replace seccomp annotations
+- Deprecate the seccomp annotations
+
+### Non-Goals
+
+This KEP proposes the absolute minimum to get seccomp to GA, therefore all functional enhancements
+are out of scope, including:
+
+- Changing the default seccomp profile from `unconfined`
+- Defining any standard "Kubernetes branded" seccomp profiles
+- Formally speccing the seccomp profile format in Kubernetes
+- Providing mechanisms for loading profiles from outside the static seccomp node directory
+- Changing the semantics around seccomp support
+- Windows support (seccomp is very linux-specific)
+
+## Proposal
+
+### API
+
+The seccomp API will be functionally equivalent to the current alpha API. This includes the Pod API,
+which specifies what profile the pod & containers run with, and the PodSecurityPolicy API which
+specifies allowed profiles & a default profile.
+
+#### Pod API
+
+The Pod Seccomp API is immutable, except in [`PodTemplates`](#podtemplates).
+
+```go
+type PodSecurityContext struct {
+    ...
+    // The seccomp options to use by the containers in this pod.
+    // +optional
+    Seccomp  *SeccompOptions
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The seccomp options to use by this container. If seccomp options are
+    // provided at both the pod & container level, the container options
+    // override the pod options.
+    // +optional
+    Seccomp  *SeccompOptions
+    ...
+}
+
+type SeccompOptions struct {
+    // The seccomp profile to run with.
+    SeccompProfile
+}
+
+// Only one profile source may be set.
+// +union
+type SeccompProfile struct {
+    // +unionDescriminator
+    Type SeccompProfileType
+    // Use a predefined profile defined by the runtime.
+    // Most runtimes only support "default"
+    // +optional
+    RuntimeProfile *string
```

Comment:
specify format limits (length, valid characters)

Reply:
As I go through this, I'm second guessing this field. AFAIK there aren't any runtimes that support multiple built in profiles, and this feature has never been requested.

If we dropped this field for now (just assume it's runtime/default), how bad would it be to add it back at some point in the future if it was needed? As long as we defaulted the new field to "default" and it's immutable, it doesn't seem like it would be that problematic? Or am I forgetting something?